pkg/cloud/cloud.go

```diff
@@ -515,12 +517,24 @@ func (c *cloud) AttachDisk(ctx context.Context, volumeID, nodeID string) (string
 		return "", err
 	}
 
+	disk, err := c.GetDiskByID(ctx, volumeID)
+	if err != nil {
+		if errors.Is(err, ErrNotFound) {
+			return "", status.Errorf(codes.NotFound, "volume %q not found", volumeID)
+		}
+		return "", fmt.Errorf("could not get volume with ID %q: %w", volumeID, err)
+	}
+
 	device, err := c.dm.NewDevice(instance, volumeID)
 	if err != nil {
-		return "", err
+		return "", fmt.Errorf("could not create device for volume %q: %w", volumeID, err)
 	}
 	defer device.Release(false)
 
+	if isDiskAttached(disk, nodeID) && device.IsAlreadyAssigned {
```

Comment:
Is this not vulnerable to the "time travel" problem documented here: https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/master/pkg/cloud/cloud.go#L680-L683 ?

Reply:
The `WaitForAttachmentState` function is polling AWS for the current state of the volume attachment and waits until the volume reaches the expected state (attached or detached). It checks the state reported by AWS directly. If AWS's eventual consistency leads to temporarily incorrect state reports, this function is designed to handle that. It retries until the state matches the expected state, or until the maximum number of retries is exhausted.

When the `AttachDisk` function checks `if isDiskAttached(disk, nodeID) && device.IsAlreadyAssigned`, it's checking the local state before calling `WaitForAttachmentState`. This local state check is "safe" in the sense that it's based on the state as understood by the cloud controller, which should be internally consistent.

However, the real concern with AWS's eventual consistency is how it impacts interactions with AWS's APIs, such as when issuing attach requests or when polling for the attachment state. These are the points where the cloud controller's understanding of the state needs to align with AWS's understanding of the state. That's why error checking and retries are used in these parts of the code.

For the specific case of the device already being assigned and the disk already being attached (as per `isDiskAttached(disk, nodeID) && device.IsAlreadyAssigned`), this should be a stable state. If AWS reports a different state due to eventual consistency issues, the `WaitForAttachmentState` function handles this by continuously polling until the reported state matches the expected state.

So, the line is safe because it relies on internal consistency within the cloud controller. Any potential inconsistencies arising from AWS's eventual consistency are handled in the parts of the code that interact with AWS's APIs directly, including the `WaitForAttachmentState` function.

Reply:
I'm worried about the case where the API reports the same (**but incorrect**) state due to eventual consistency issues, as described here: https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/master/pkg/cloud/cloud.go#L680-L683